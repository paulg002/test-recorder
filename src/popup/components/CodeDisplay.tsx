import * as React from 'react';
import CodeBlock from './CodeBlock';
import type { Block } from '../../types';

export interface CodeDisplayProps {
  codeBlocks: Block[],
  destroyBlock: (index: number) => void,
  moveBlock: (dragIdx: number, dropIdx: number) => void,
}

export default ({ codeBlocks, destroyBlock, moveBlock }: CodeDisplayProps) => {
  const [dragOriginIdx, setDragOriginIdx] = React.useState<number>(-1);
  const [draggedIdx, setDraggedIdx] = React.useState<number>(-1);
  const [modifiedCodeBlocks, setModifiedCodeBlocks] = React.useState<Block[]>([]);

  React.useEffect(() => {
    setModifiedCodeBlocks([...codeBlocks]);
  }, [codeBlocks]);

  const swapBlocks = (i: number, j: number) => {
    const temp = [...modifiedCodeBlocks];
    const holder = temp[i];
    temp[i] = temp[j];
    temp[j] = holder;
    setModifiedCodeBlocks(temp);
  };

  const onDragStart = (e: React.DragEvent, i: number) => {
    setDraggedIdx(i);
    setDragOriginIdx(i);
  };

  const onDragOver = (e: React.DragEvent, i: number) => {
    e.preventDefault();
    if (draggedIdx !== i) {
      swapBlocks(draggedIdx, i);
      setDraggedIdx(i);
    }
  };

  const onDragEnd = () => {
    setDraggedIdx(-1);
    setModifiedCodeBlocks([...codeBlocks]);
  };

  const onDrop = (e: React.DragEvent, i: number) => {
    e.preventDefault();
    if (dragOriginIdx !== i) moveBlock(dragOriginIdx, i);
    setDragOriginIdx(-1);
  };

  const blocks = modifiedCodeBlocks.map((block, idx) => (
    <CodeBlock
      key={block.id}
      dragStatus={idx === draggedIdx ? 'drag-origin' : 'code-block'}
      text={block.value}
      index={idx}
      destroyBlock={destroyBlock}
      onDragStart={onDragStart}
      onDragOver={onDragOver}
      onDragEnd={onDragEnd}
      onDrop={onDrop}
    />
  ));

  return (
    <ul id="code-display">
      {blocks}
    </ul>
  );
};
