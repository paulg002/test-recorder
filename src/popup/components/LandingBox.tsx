import * as React from 'react';
import CypressCone from './CypressCone';

export interface LandingBoxProps { isValidTab: boolean }

export default ({ isValidTab }: LandingBoxProps) => (
  <div id="landing-box">
    <CypressCone />
    {(
      isValidTab
        ? <p className="text">Click &apos;Start Recording&apos; to start recording!</p>
        : <p className="text">This is not a valid page for recording.</p>
    )}
  </div>
);
