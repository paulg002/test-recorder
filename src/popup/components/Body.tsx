import * as React from 'react';
import type { Block } from '../../types';
import { RecState } from '../../constants';
import CodeDisplay from './CodeDisplay';
import LandingBox from './LandingBox';

export interface BodyProps {
  isValidTab: boolean,
  recStatus: RecState,
  codeBlocks: Block[],
  destroyBlock: (index: number) => void,
  moveBlock: (dragIdx: number, dropIdx: number) => void,
}

export default ({
  recStatus,
  codeBlocks,
  isValidTab,
  destroyBlock,
  moveBlock,
}: BodyProps) => (
  <div id="body">
    {recStatus === RecState.OFF && <LandingBox isValidTab={isValidTab} />}
    {recStatus !== RecState.OFF && <CodeDisplay codeBlocks={codeBlocks} destroyBlock={destroyBlock} moveBlock={moveBlock} />}
  </div>
);
