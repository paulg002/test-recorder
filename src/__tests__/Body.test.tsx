import * as React from 'react';
import { shallow } from 'enzyme';
import Body, { BodyProps } from '../popup/components/Body';
import LandingBox from '../popup/components/LandingBox';
import CodeDisplay from '../popup/components/CodeDisplay';
import { RecState } from '../constants';
import '../setupTests';

describe('Body', () => {
  let props: BodyProps;
  let wrapper;
  beforeAll(() => {
    props = {
      codeBlocks: [],
      isValidTab: true,
      destroyBlock: jest.fn(),
      moveBlock: jest.fn(),
      recStatus: RecState.OFF,
    };
  });
  it('Should render LandingBox when recStatus is off', () => {
    wrapper = shallow(<Body {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.exists(LandingBox)).toBe(true);
    expect(wrapper.exists(CodeDisplay)).toBe(false);
  });
  it('Should render CodeDisplay when recStatus is paused', () => {
    props.recStatus = RecState.PAUSED;
    wrapper = shallow(<Body {...props} />);
    expect(wrapper.exists(CodeDisplay)).toBe(true);
    expect(wrapper.exists(LandingBox)).toBe(false);
  });
  it('Should render CodeDisplay when recStatus is on', () => {
    props.recStatus = RecState.ON;
    wrapper = shallow(<Body {...props} />);
    expect(wrapper.exists(CodeDisplay)).toBe(true);
    expect(wrapper.exists(LandingBox)).toBe(false);
  });
});
