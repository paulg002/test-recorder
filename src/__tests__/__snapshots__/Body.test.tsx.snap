// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body Should render LandingBox when recStatus is off 1`] = `
<div
  id="body"
>
  <Component
    isValidTab={true}
  />
</div>
`;
